js/ui/windowManager.js

```diff
@@ -1832,3 +1887,608 @@ const WindowManager = new Lang.Class({
+        button.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        Main.layoutManager.addChrome(button);
+        button.connect('clicked', Lang.bind(this, this._switchToApp, session));
+
+        session.positionChangedIdBuilder = window.connect('position-changed', Lang.bind(this, this._windowBuilderPositionChanged, session));
+        session.sizeChangedIdBuilder = window.connect('size-changed', Lang.bind(this, this._windowBuilderSizeChanged, session));
+
+        session.buttonBuilder = button;
+    },
+
+    _animateToBuilder: function(session) {
+        /* We wait until the first frame of the window has been drawn
+         * and damage updated in the compositor before we start rotating.
+         *
+         * This way we don't get ugly artifacts when rotating if
+         * a window is slow to draw. */
+        let firstFrameConnection = session.actorBuilder.connect('first-frame', Lang.bind(this, function() {
+            // reset the bouncing animation that was showed while Builder was loading
+            Tweener.removeTweens(session.buttonApp);
+            session.buttonApp.scale_y = 1;
+            session.buttonApp.scale_x = 1;
+            session.buttonApp.translation_y = 0;
+            session.buttonApp.translation_x = 0;
+
+            this._animate(session.actorApp, session.actorBuilder, Gtk.DirectionType.LEFT);
+
+            this._firstFrameConnections = this._firstFrameConnections.filter(function(conn) {
+                return conn != session.firstFrameConnection;
+            });
+            session.actorBuilder.disconnect(session.firstFrameConnection);
+            session.firstFrameConnection = null;
+
+            this._addButton(session);
+            session.buttonBuilder.show();
+            session.buttonApp.hide();
+
+            return false;
+        }));
+
+        this._prepareAnimate(session.actorApp, session.actorBuilder, Gtk.DirectionType.LEFT);
+        /* Save the connection's id in the session and in a list too so we can
+         * get rid of it on kill-window-effects later */
+        session.firstFrameConnection = firstFrameConnection;
+        this._firstFrameConnections.push(firstFrameConnection);
+    },
+
+    _windowAppPositionChanged: function(window, session) {
+        let rect = session.actorApp.meta_window.get_frame_rect();
+        session.buttonApp.set_position(rect.x + rect.width - BUTTON_OFFSET_X, rect.y + rect.height - BUTTON_OFFSET_Y);
+        if (!session.actorBuilder)
+            return;
+        session.actorBuilder.meta_window.move_resize_frame(false,
+                                                           rect.x,
+                                                           rect.y,
+                                                           rect.width,
+                                                           rect.height);
+    },
+
+    _windowAppSizeChanged: function(window, session) {
```

Comment:
This function does exactly what `_windowAppPositionChanged()` does above. Can you factor both into an `_updateAppSizeAndPosition()`?